x/sp/keeper/sp_status.go

```diff
@@ -0,0 +1,106 @@
+	k.cdc.MustUnmarshal(bz, &stats)
+	size := len(stats.Records)
+	totalUsedTime := int64(0)
+	for i := size - 1; i >= 0; i-- {
+		record := stats.Records[i]
+		if i == size-1 {
+			threshHold := record.GetHeight() + params.GetNumOfLockupBlocksForMaintenance()
+			if ctx.BlockHeight() < threshHold {
+				return errors.Wrapf(types.ErrStorageProviderStatusUpdateNotAllow, "wait after block height %d", threshHold)
+			}
+		}
+		totalUsedTime = totalUsedTime + record.GetActualDuration()
+	}
+	if totalUsedTime+requestDuration > params.GetMaintenanceDurationQuota() {
+		return errors.Wrapf(types.ErrStorageProviderStatusUpdateNotAllow, "not enough quota, quota=%d is less than requested=%d", params.GetMaintenanceDurationQuota()-totalUsedTime, requestDuration)
+	}
+	stats.Records = append(stats.Records, newRecord)
+	store.Set(key, k.cdc.MustMarshal(&stats))
+	return nil
+}
+
+func (k Keeper) UpdateToInService(ctx sdk.Context, sp *types.StorageProvider) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.GetStorageProviderMaintenanceRecordsKey(sdk.MustAccAddressFromHex(sp.OperatorAddress))
+	bz := store.Get(key)
+	if bz != nil {
+		var stats types.SpMaintenanceStats
+		k.cdc.MustUnmarshal(bz, &stats)
+		size := len(stats.Records)
+		if size != 0 {
+			lastRecord := stats.Records[size-1]
+			lastRecord.ActualDuration = ctx.BlockTime().Unix() - lastRecord.RequestAt
+			store.Set(key, k.cdc.MustMarshal(&stats))
+		}
+	}
+	sp.Status = types.STATUS_IN_SERVICE
+}
+
+func (k Keeper) ForceUpdateMaintenanceRecords(ctx sdk.Context) {
+	params := k.GetParams(ctx)
+	store := ctx.KVStore(k.storeKey)
+	curTime := ctx.BlockTime().Unix()
+	iter := storetypes.KVStorePrefixIterator(store, types.StorageProviderKey)
+	for ; iter.Valid(); iter.Next() {
+		sp := types.MustUnmarshalStorageProvider(k.cdc, iter.Value())
+		key := types.GetStorageProviderMaintenanceRecordsKey(sdk.MustAccAddressFromHex(sp.OperatorAddress))
+		bz := store.Get(key)
+		if bz != nil {
+			var stats types.SpMaintenanceStats
+			k.cdc.MustUnmarshal(bz, &stats)
+			size := len(stats.Records)
+			// force update any maintenance record that not been updated back to in_service after requested duration.
+			if sp.Status != types.STATUS_IN_SERVICE {
+				for i := size - 1; i >= 0; i-- {
+					if stats.Records[i].GetActualDuration() == 0 && stats.Records[i].RequestAt+stats.Records[i].GetRequestDuration() < curTime {
+						stats.Records[i].ActualDuration = stats.Records[i].RequestDuration
+						store.Set(key, k.cdc.MustMarshal(&stats))
+						sp.Status = types.STATUS_IN_SERVICE
+						k.SetStorageProvider(ctx, sp)
+					}
+				}
+			}
+			// purge outdated records
+			for i := size - 1; i >= 0; i-- {
+				if stats.Records[i].GetHeight()+params.GetNumOfHistoricalBlocksForMaintenanceRecords() < ctx.BlockHeight() {
+					stats.Records = stats.Records[i+1:]
+					break
+				}
+			}
+			if len(stats.Records) == 0 {
+				store.Delete(key)
+			} else {
+				store.Set(key, k.cdc.MustMarshal(&stats))
```

Comment:
I think in most case, there is no need to set here, can we have an indicator to know whether we should set or not.